blog/2022/04/simplechains.md

```diff
@@ -0,0 +1,489 @@
+
+## Scientific Machine Learning (SciML) and "Small" Neural Networks
+
+SimpleChains.jl is a library developed by Julia Computing and PumasAI in collaboration with Roche and the University of Maryland, Baltimore. The purpose of SimpleChains.jl is to be as fast as possible for small neural networks. This is because SimpleChains.jl originated as a solution for the DeepPumas product for scientific machine learning (SciML) in clinical pharmacology. In this domain, small neural networks are combined with known nonlinear mixed effects models (statistical models with differential equations) to discover previously unknown mechanisms and prognostic factors. For a short introduction to how this is done, check out the following video by Niklas Korsbo:
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/TFB_lt1KMto" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+
+This SciML methodology has been shown across many disciplines, from black hole dynamics to the development of earthquake safe buildings, to be a flexible method capable of discovering guiding (bio)physical equations. Here's a recent talk with walks through the various use cases of SciML throughout the sciences:
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/eSeY4K4bITI?start=668" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+
+For more details on the software and methods, check out this manuscript. 
+
+However, the unique aspects of how neural networks are used in these contexts make it rife for specializing of performance. Specifically, in the context of machine learning one normally relies on the following assumption: the neural networks are large enough that the O(n^3) cost of matrix-matrix multiplication (or other kernels like convolutions) domainates the the runtime. This is essentially the guiding principle behind most of the mechanics of a machine learning library:
+
+1. Matrix-matrix multiplication scales cubicly while memory allocations scale linearly, so attempting to mutate vectors with non-allocating operations is not a high priority. Just use `A*x`.
+2. Focus on accelearting GPU kernels as fast as possible! Since these large matrix-matrix operations will be fastest on GPUs and are the bottleneck, performance benchmarks will essentially just be a measurement of how fast these specific kernels are.
+3. When doing reverse-mode automatic differentiation (backpropogation), feel free to copy values to memory. Memory allocations will be hidden by the larger kernel calls. 
+4. Also, feel free to write a "tape" for generation backpropogation. The tape does add the cost of essentially building a dictionary during the forward pass, but that will be hidden by the larger kernel calls.
```

Comment:
```suggestion
4. Also, feel free to write a "tape" for generating backpropogation. The tape does add the cost of essentially building a dictionary during the forward pass, but that will be hidden by the larger kernel calls.
```
Also, optimizing activation functions isn't a high priority, because (like copying) that scales linearly.
SimpleChains will fuse certain cheap activation functions with matmul and conv kernels, e.g. `relu`.
Unlike Flux, it'll SIMD all of them.

Reply:
yeah add it?